### PR TITLE
Property and Method strategies

### PR DESCRIPTION
**Pull request by jaapio**

### CODE REVIEW

File: src/phpDocumentor/Reflection/Php/Factory/Method.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        if (!$this->matches($object)) {
+            throw new InvalidArgumentException(
+                sprintf('%s cannot handle objects with the type %s',
+                    __CLASS__,
+                    is_object($object) ? get_class($object) : gettype($object)
+                )
+            );
+        }
+
+        $method = new MethodDescriptor(new Fqsen($object->name));
```
Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/Method.php]: I am missing the determination and passing whether a method is final, abstract, static and what visibility it has

Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/Method.php]: Is $object->name a valid Fqsen for this element? I would expect it only contains the name of the method without the `()` parenthesis while an Fqsen has the class name, double colon separator and method name including parenthesis

Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/Method.php]: Do I guess correctly that you will rename/move the MethodDescriptor later?

File: src/phpDocumentor/Reflection/Php/Factory/Property.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        if (!$this->matches($object)) {
+            throw new InvalidArgumentException(
+                sprintf('%s cannot handle objects with the type %s',
+                    __CLASS__,
+                    is_object($object) ? get_class($object) : gettype($object)
+                )
+            );
+        }
+
+        $visibility = new Visibility($object->visibility);
```
Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/Property.php]: The PropertyNode does not have a property visibility; it has a property type that contains a bitflag determining the visibility and whether a property set it static.

Please also note that the PropertyNode does not equal one property, the following is one PropertyNode
`public static $property1, $property2;`

See the props property in the PropertyNode for PropertyProperty nodes (one or more); it is the PropertyProperty that contains the property name (without dollar sign and class) and default value

File: src/phpDocumentor/Reflection/Php/Factory/Property.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+        return new PropertyDescriptor(new Fqsen($object->name), $visibility, null, $object->default, $object->static);
```
Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/Property.php]: As indicated above; $object has no name, default and no static property.

Also: a property mentioned in the props property its name is not a Fqsen but only the name of the property itself without dollar sign

File: src/phpDocumentor/Reflection/Php/Factory/Method.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+        foreach ($object->params as $param) {
+            $strategy = $strategies->findMatching($param);
+            $method->addArgument($strategy->create($param, $strategies));
+        }
+
+        return $method;
+    }
+}
```
Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/Method.php]: Missing newline

**Comment by mvriel** [comment]
Cannot merge this PR as it contains a few points that need to be addresses; I can recommend examining the original MethodAssembler and PropertyAssembler factories to see how they approached the situation

File: src/phpDocumentor/Reflection/Php/Factory/Method.php (hunk 1 above)
Comment (jaapio) [inline src/phpDocumentor/Reflection/Php/Factory/Method.php]: MethodDescriptor is an namespace alias ;-)

**Comment by jaapio** [comment]
I think this one can be checked again and merged.

File: src/phpDocumentor/Reflection/Php/Factory/Method.php [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            throw new InvalidArgumentException(
+                sprintf('%s cannot handle objects with the type %s',
+                    __CLASS__,
+                    is_object($object) ? get_class($object) : gettype($object)
+                )
+            );
+        }
+
+        $method = new MethodDescriptor(
+            new Fqsen($object->name . '()'),
```
Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/Method.php]: The name property is only a local name while Fqsen is the complete name including namespace and class; so with the method `myMethod` in the class `MyNamespace\MyClass` will `$object->name` return `myMethod` while what you want is `\MyNamespace\MyClass::myMethod()`

File: src/phpDocumentor/Reflection/Php/Factory/Property.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+                sprintf('%s cannot handle objects with the type %s',
+                    __CLASS__,
+                    is_object($object) ? get_class($object) : gettype($object)
+                )
+            );
+        }
+
+        $visibility = $this->buildVisibility($object);
+
+        return new PropertyDescriptor(new Fqsen($object->name), $visibility, null, $object->default, $object->static);
```
Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/Property.php]: The name property is only a local name while Fqsen is the complete name including namespace and class; so with the method `myMethod` in the class `MyNamespace\MyClass` will `$object->name` return `myMethod` while what you want is `\MyNamespace\MyClass::myMethod()`

Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/Property.php]: `$object->static` does not exist, you probably mean `$object->isStatic()`?

Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/Property.php]: $object->default is either null or the class `Node\Expr`; but the $default parameter should  only accept strings

Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/Property.php]: @jaapio you can use the PrettyPrinter to convert a Node\Expr to string; see https://github.com/phpDocumentor/Reflection/blob/develop/src/phpDocumentor/Descriptor/Builder/PhpParser/AssemblerAbstract.php#L202 for an example in the current codebase.

Please note that phpDocumentor uses an adjusted prettyprinter to address an edge case

**Pull request merged**